umbra-js/test/Umbra.test.ts

```diff
@@ -176,172 +206,248 @@ describe('Umbra class', () => {
+    describe(`${test.name}, scan, and withdraw funds`, () => {
+      beforeEach(() => {
+        // Seems we somehow lose the provider attached to our sender, so make sure it's there. Without this
+        // some tests below throw with "Error: missing provider (operation="sendTransaction", code=UNSUPPORTED_OPERATION, version=abstract-signer/5.0.12)"
+        sender = sender.connect(ethers.provider);
+      });
+
+      const mintAndApproveDai = async (signer: Wallet, user: string, amount: BigNumber) => {
+        await dai.connect(signer).mint(user, amount);
+        await dai.connect(signer).approve(umbra.umbraContract.address, ethers.constants.MaxUint256);
+        await dai.connect(signer).approve(umbra.batchSendContract.address, ethers.constants.MaxUint256);
+      };
+
+      it('reverts if sender does not have enough tokens', async () => {
+        const msg = `Insufficient balance to complete transfer. Has 0 tokens, tried to send ${quantity.toString()} tokens.`;
+
+        if (test.id === 'send') {
+          await expectRejection(umbra.send(sender, dai.address, quantity, receiver.address), msg);
+        } else if (test.id === 'batchSend') {
+          await expectRejection(
+            umbra.batchSend(
+              sender,
+              [{ token: dai.address, amount: quantity, address: receiver!.publicKey }],
+              overrides
+            ),
+            msg
+          );
+        }
+      });
+
+      it('reverts if sender does not have enough ETH', async () => {
+        // ETH balance is checked by ethers when sending a transaction and therefore does not need to
+        // be tested here. If the user has insufficient balance it will throw with
+        // `insufficient funds for gas * price + value`
+      });
+
+      it('Send tokens, scan for them, withdraw them (direct withdraw)', async () => {
+        // SENDER
+        // Mint Dai to sender, and approve the Umbra contract to spend their DAI
+        await mintAndApproveDai(sender, sender.address, quantity.mul(5));
+
+        // Send funds with Umbra
+        let tx: ContractTransaction | null = null;
+        let stealthKeyPairs: KeyPair[] = [];
+        let usedReceivers: Wallet[] = [];
+
+        if (test.id === 'send') {
+          const result = await umbra.send(sender, dai.address, quantity, receiver!.publicKey, overrides);
+          tx = result.tx;
+          stealthKeyPairs = [result.stealthKeyPair];
+          usedReceivers = receivers.slice(0, 1);
+        } else if (test.id === 'batchSend') {
+          const sends: SendBatch[] = [];
+          for (let i = 0; i < 5; i++) {
+            sends.push({ token: dai.address, amount: quantity, address: receivers[i].publicKey });
+          }
+          const result = await umbra.batchSend(sender, sends, overrides);
+          tx = result.tx;
+          stealthKeyPairs = result.stealthKeyPairs;
+          usedReceivers = receivers.slice(0, 5);
+        }
+        if (tx) await tx.wait();
```

Comment:
We know `tx` will always be defined, so semantically this is more accurate

```suggestion
        await tx!.wait();
```

Reply:
Also here: https://github.com/ScopeLift/umbra-protocol/pull/426/commits/3d3234a5ad319ef3bfa08cc8d8951239baa963d2